### PR TITLE
#d2l  12.1 12.2 12.3 12.5 12.6 14.8 14.10

### PR DESCRIPTION
**Pull request by skywalk163**
动手学飞桨版本12.1节

需要同步的paddle.py文件内容如下：

```
#@save
class Benchmark:
    """用于测量运行时间"""
    def __init__(self, description='Done'):
        self.description = description

    def __enter__(self):
        self.timer = d2l.Timer()
        return self

    def __exit__(self, *args):
        print(f'{self.description}: {self.timer.stop():.4f} sec')
```

### CODE REVIEW

**Comment by lvjian0706** [comment]
12.1：
1.“通过混合式编程加速”：时间测试结果与结论不符，应该是静态图时间更短，可以确认下代码是否有问题
12.2:
1.“通过后端异步处理”：文档是pytorch，可以看下能否调整成paddle的说明
2.“障碍器与阻碍器”、“改进计算”：这部分没有写？
12.5：
1.“数据并行性”：文档中“在实践中请注意，当在k个GPU上训练时，需要扩大小批量的大小为k的倍数”在paddle中不适用
2.“训练”中epoch设置与原版不统一
12.6:
1.“网络初始化”：建议与原版保持一致，否则这一部分没有出现的必要
2.原版使用的是d2l的评估代码，这里为什么直接调用的paddle？
3.训练时长没有计算和打印
4.2GPU时，为什么重新实现了一遍代码？

**Comment by skywalk163** [comment]
```
12.1：
1.“通过混合式编程加速”：时间测试结果与结论不符，应该是静态图时间更短，可以确认下代码是否有问题。 
答：会再次确认。
12.2:
1.“通过后端异步处理”：文档是pytorch，可以看下能否调整成paddle的说明 。 
答：前期是讲的不调文档。这里可以做相应调整。
2.“障碍器与阻碍器”、“改进计算”：这部分没有写？
答：因为前期沟通的是修改torch部分，这部分torch没有写，所以才没写。请确认这部分是否需要加上。
12.5：
1.“数据并行性”：文档中“在实践中请注意，当在k个GPU上训练时，需要扩大小批量的大小为k的倍数”在paddle中不适用
答：会进行修改。
2.“训练”中epoch设置与原版不统一
答：会修改成跟原版一样。
12.6:
1.“网络初始化”：建议与原版保持一致，否则这一部分没有出现的必要
答：我试试能否与原版保持一致，这里有点难度，因为飞桨用的是fleet实现，要在模型定义前初始化，总体感觉结构肯定要改。 或者这里是用spawn实现，非常简单方便，前期没有用spawn实现是因为我在使用spawn的时候在双p80报错，经过沟通、调试，后在双T4和V100下调通。（大约是某平台的双P80的锅）。（但是spawn实现也较难跟原版一致。） 
后来看了一下，torch也没有在“网络初始化“这一段进行初始化，而是到训练里面去初始化的。也就是torch为了跟mxnet对齐保留了这一节，但没有初始化。对飞桨来说，这节同样没意义。
2.原版使用的是d2l的评估代码，这里为什么直接调用的paddle？
答：这里我再去调试一下，我记得是因为飞桨的多GPU简洁实现必须在命令行下实现，这样d2l的图形实现无法显示而失效，所以才添加了飞桨自己的。
3.训练时长没有计算和打印
答：这块要加上。
4.2GPU时，为什么重新实现了一遍代码？
答：是为了将代码写入一个文件，然后使用python xx.py 或者 python -m paddle.distributed.launch xx.py  来双卡执行。
```

**Comment by tngt** [comment]
12.2:
1.“通过后端异步处理”：文档是pytorch，可以看下能否调整成paddle的说明 。 
答：前期是讲的不调文档。这里可以做相应调整。  

**涉及到飞桨相关的理论，段老师可以评估下是否好写，如果对该方面不了解，可以反馈给我。**
------------------------------------------------------------------------------------------------------
2.“障碍器与阻碍器”、“改进计算”：这部分没有写？
答：因为前期沟通的是修改torch部分，这部分torch没有写，所以才没写。请确认这部分是否需要加上。

**该部分可以先尝试参考MXNET尝试实现，如果实现困难，可以暂时不写。**

**Comment by skywalk163** [comment]
12.1：
1.“通过混合式编程加速”：时间测试结果与结论不符，应该是静态图时间更短，可以确认下代码是否有问题。 
答：会再次确认。个人认为是因为小模型，并且静态图最后结果会回传到动态图导致的消耗所致，报issue咨询https://github.com/PaddlePaddle/Paddle/issues/39794
12.2:
1.“通过后端异步处理”：文档是pytorch，可以看下能否调整成paddle的说明 。 
答：这里已经做了相应修改。当然文档还需要专业的老师再精细打磨下。
2.“障碍器与阻碍器”、“改进计算”：这部分没有写？
答：这部分已经加上。
12.5：
1.“数据并行性”：文档中“在实践中请注意，当在k个GPU上训练时，需要扩大小批量的大小为k的倍数”在paddle中不适用
答：已经进行修改。
2.“训练”中epoch设置与原版不统一
答：已修改成跟原版一样。

**Comment by skywalk163** [comment]
12.6章节进行了一些调整

1.“网络初始化”：建议与原版保持一致，否则这一部分没有出现的必要
答：还是放入训练部分了。
2.原版使用的是d2l的评估代码，这里为什么直接调用的paddle？
答：这里既用了d2l的评估代码，也使用了飞桨自己了，是因为飞桨的多GPU简洁实现必须在命令行下实现，这样d2l的图形实现无法显示而失效，所以才添加了飞桨自己的。
3.训练时长没有计算和打印
答：这块已加上。

代码重新修改过了，在双卡系统验证通过。

**Comment by lvjian0706** [comment]
12.1中静态图代码问题，可以参考下面链接进行再次修改：
1.在静态图模式下，需要创建program
2.必需paddle.enable_static()才能运行静态图代码
可以参考：
https://github.com/PaddlePaddle/PaddleClas/blob/1bec0806f67af03dcf42c176a4ac79a1daca8e88/ppcls/static/train.py#L129
https://www.paddlepaddle.org.cn/documentation/docs/zh/api/paddle/static/Program_cn.html#cn-api-fluid-program

**Comment by lvjian0706** [comment]
12.6 多GPU训练中，不需要使用‘fleet’，可以参考：https://www.paddlepaddle.org.cn/documentation/docs/zh/guides/02_paddle2.0_develop/06_device_cn.html

**Comment by skywalk163** [comment]
> 12.1中静态图代码问题，可以参考下面链接进行再次修改： 1.在静态图模式下，需要创建program 2.必需paddle.enable_static()才能运行静态图代码 可以参考： https://github.com/PaddlePaddle/PaddleClas/blob/1bec0806f67af03dcf42c176a4ac79a1daca8e88/ppcls/static/train.py#L129 https://www.paddlepaddle.org.cn/documentation/docs/zh/api/paddle/static/Program_cn.html#cn-api-fluid-program

12.1是在动态图下使用to_static进行静态化，不能直接用静态图哦，见动手学深度学习的原文档：https://zh.d2l.ai/chapter_computational-performance/hybridize.html

**Comment by skywalk163** [comment]
> 12.6 多GPU训练中，不需要使用‘fleet’，可以参考：https://www.paddlepaddle.org.cn/documentation/docs/zh/guides/02_paddle2.0_develop/06_device_cn.html

已修改为使用spawn并行训练。

File: Dive-into-DL-paddlepaddle/docs/14_chapter_natural-language-processing-pretraining/14.10.md [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+    return mlm_l, nsp_l, l
+```
+
+通过调用上述两个辅助函数，下面的`train_bert`函数定义了在WikiText-2（`train_iter`）数据集上预训练BERT（`net`）的过程。训练BERT可能需要很长时间。以下函数的输入`num_steps`指定了训练的迭代步数，而不是像`train_ch13`函数那样指定训练的轮数（参见 :numref:`sec_image_augmentation`）。
+
+
+
+```python
+def train_bert(train_iter, net, loss, vocab_size, devices, num_steps):
+    trainer = paddle.optimizer.Adam(parameters=net.parameters(), learning_rate=0.01)
```
Comment (w5688414) [inline Dive-into-DL-paddlepaddle/docs/14_chapter_natural-language-processing-pretraining/14.10.md:74]: 缺少一行代码,对应的torch的实现是：
```
net = nn.DataParallel(net, device_ids=devices).to(devices[0])
```

File: Dive-into-DL-paddlepaddle/docs/14_chapter_natural-language-processing-pretraining/14.10.md [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+    step, timer = 0, d2l.Timer()
+    animator = d2l.Animator(xlabel='step', ylabel='loss',
+                            xlim=[1, num_steps], legend=['mlm', 'nsp'])
+    # 遮蔽语言模型损失的和，下一句预测任务损失的和，句子对的数量，计数
+    metric = d2l.Accumulator(4)
+    num_steps_reached = False
+    while step < num_steps and not num_steps_reached:
+        for tokens_X, segments_X, valid_lens_x, pred_positions_X,\
+            mlm_weights_X, mlm_Y, nsp_y in train_iter:
+            timer.start()
```
Comment (w5688414) [inline Dive-into-DL-paddlepaddle/docs/14_chapter_natural-language-processing-pretraining/14.10.md:85]: trainer.clear_grad()放在这里，跟pytorch保持一致

File: Dive-into-DL-paddlepaddle/docs/14_chapter_natural-language-processing-pretraining/14.10.md [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+```
+
+## 用BERT表示文本
+
+在预训练BERT之后，我们可以用它来表示单个文本、文本对或其中的任何词元。下面的函数返回`tokens_a`和`tokens_b`中所有词元的BERT（`net`）表示。
+
+
+
+```python
+
```
Comment (w5688414) [inline Dive-into-DL-paddlepaddle/docs/14_chapter_natural-language-processing-pretraining/14.10.md:124]: 空行删除

File: Dive-into-DL-paddlepaddle/docs/14_chapter_natural-language-processing-pretraining/14.10.md [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+
+def get_bert_encoding(net, tokens_a, tokens_b=None):
+    tokens, segments = d2l.get_tokens_and_segments(tokens_a, tokens_b)
+    token_ids = paddle.to_tensor(vocab[tokens]).unsqueeze(0)
+    segments = paddle.to_tensor(segments).unsqueeze(0)
+    valid_len = paddle.to_tensor(len(tokens))
+
```
Comment (w5688414) [inline Dive-into-DL-paddlepaddle/docs/14_chapter_natural-language-processing-pretraining/14.10.md:131]: 空行删除

File: Dive-into-DL-paddlepaddle/docs/14_chapter_natural-language-processing-pretraining/14.10.md [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+```
+
+在预训练过程中，我们可以绘制出遮蔽语言模型损失和下一句预测损失。
+
+
+
+```python
+train_bert(train_iter, net, loss, len(vocab), devices, 10)
+
```
Comment (w5688414) [inline Dive-into-DL-paddlepaddle/docs/14_chapter_natural-language-processing-pretraining/14.10.md:114]: 删除冗余的行

Comment (w5688414) [inline Dive-into-DL-paddlepaddle/docs/14_chapter_natural-language-processing-pretraining/14.10.md:114]: 测试的paddle的结果是：

```
MLM loss 7.111, NSP loss 1.472
```
pytorch的结果是：

```
MLM loss 5.639, NSP loss 0.758
```
看能不能调得跟pytorch相当

File: Dive-into-DL-paddlepaddle/docs/14_chapter_natural-language-processing-pretraining/14.10.md (hunk 1 above)
Comment (skywalk163) [reply Dive-into-DL-paddlepaddle/docs/14_chapter_natural-language-processing-pretraining/14.10.md:74]: > DataParallel

这句负责数据并行的，在飞桨这边如果要使用这句话的话，需要配合paddle.distributed ,使用launch或者spawn方式。
如果只写这一句话，也不是不可以，请再确认一下。

**Comment by lvjian0706** [comment]
> > 12.6 多GPU训练中，不需要使用‘fleet’，可以参考：https://www.paddlepaddle.org.cn/documentation/docs/zh/guides/02_paddle2.0_develop/06_device_cn.html
> 
> 已修改为使用spawn并行训练。

1. epoch和lr没有对齐
2. 建议改为launch启动

File: Dive-into-DL-paddlepaddle/docs/14_chapter_natural-language-processing-pretraining/14.10.md (hunk 5 above)
Comment (skywalk163) [reply Dive-into-DL-paddlepaddle/docs/14_chapter_natural-language-processing-pretraining/14.10.md:114]: 我再尝试一下。

**Comment by skywalk163** [comment]
> > > 12.6 多GPU训练中，不需要使用‘fleet’，可以参考：https://www.paddlepaddle.org.cn/documentation/docs/zh/guides/02_paddle2.0_develop/06_device_cn.html
> > 
> > 
> > 已修改为使用spawn并行训练。
> 
> 1. epoch和lr没有对齐
> 2. 建议改为launch启动

新提交
epoch和lr对齐，
改为launch启动。

File: Dive-into-DL-paddlepaddle/docs/14_chapter_natural-language-processing-pretraining/14.10.md (hunk 5 above)
Comment (skywalk163) [reply Dive-into-DL-paddlepaddle/docs/14_chapter_natural-language-processing-pretraining/14.10.md]: 最新结果 MLM loss 5.781, NSP loss 0.842

**Comment by w5688414** [comment]
14.8 节，LGTM

**Comment by w5688414** [comment]
14.10 LGTM

**Comment by tngt** [comment]
review全部通过，可以merge
